src/library/scala/collection/immutable/ListMap.scala

```diff
@@ -32,13 +33,45 @@ object ListMap extends ImmutableMapFactory[ListMap] {
     new MapCanBuildFrom[A, B]
   def empty[A, B]: ListMap[A, B] = EmptyListMap.asInstanceOf[ListMap[A, B]]
 
+  override def newBuilder[A, B]: Builder[(A, B), ListMap[A, B]] = new ListMapBuilder[A, B]
+
   @SerialVersionUID(-8256686706655863282L)
   private object EmptyListMap extends ListMap[Any, Nothing] {
     override def apply(key: Any) = throw new NoSuchElementException("key not found: " + key)
     override def contains(key: Any) = false
     override def last: (Any, Nothing) = throw new NoSuchElementException("Empty ListMap")
     override def init: ListMap[Any, Nothing] = throw new NoSuchElementException("Empty ListMap")
   }
+
+  /**
+    * A custom builder because forgetfully adding elements one at a time to a list backed map puts
+    * the "squared" in N^2. There is a temporary space cost, but it's improbable a list backed set
+    * could become large enough for this to matter given its pricy element lookup.
+    *
+    * This builder is reusable.
+    **/
+  class ListMapBuilder[A, B](initial: ListMap[A, B]) extends ReusableBuilder[(A, B), ListMap[A, B]] {
+    def this() = this(empty[A, B])
+    protected val elems = (new mutable.ListBuffer[(A, B)] ++= initial).reverse
+    protected val seen = new mutable.HashMap[A, B] ++= initial
```

Comment:
Uh--this could get super, super expensive with small maps, which arguably is what everyone ought to have been using ListMap for to begin with.  Have you benchmarked how bad this is for lists of size 3 or so?